// webcrypto-core/build/aes/crypto.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var error_1 = require("../error");
var base_1 = require("../base");
var alg_1 = require("../alg");
var AesKeyGenParamsError = (function (_super) {
    __extends(AesKeyGenParamsError, _super);
    function AesKeyGenParamsError() {
        _super.apply(this, arguments);
        this.code = 7;
    }
    return AesKeyGenParamsError;
}(error_1.AlgorithmError));
var Aes = (function (_super) {
    __extends(Aes, _super);
    function Aes() {
        _super.apply(this, arguments);
    }
    Aes.checkAlgorithm = function (alg) {
        if (alg.name.toUpperCase() !== this.ALG_NAME.toUpperCase())
            throw new error_1.AlgorithmError(error_1.AlgorithmError.WRONG_ALG_NAME, alg.name, this.ALG_NAME);
    };
    Aes.checkKeyGenParams = function (alg) {
        switch (alg.length) {
            case 128:
            case 192:
            case 256:
                break;
            default:
                throw new AesKeyGenParamsError(AesKeyGenParamsError.PARAM_WRONG_VALUE, "length", "128, 192 or 256");
        }
    };
    Aes.checkKeyGenUsages = function (keyUsages) {
        var _this = this;
        this.checkKeyUsages(keyUsages);
        keyUsages.forEach(function (usage) {
            var i = 0;
            for (i; i < _this.KEY_USAGES.length; i++)
                if (_this.KEY_USAGES[i].toLowerCase() === usage.toLowerCase()) {
                    break;
                }
            if (i === _this.KEY_USAGES.length)
                throw new error_1.WebCryptoError("Unsuported key usage '" + usage + "'. Should be one of [" + _this.KEY_USAGES.join(", ") + "]");
        });
    };
    Aes.generateKey = function (algorithm, extractable, keyUsages) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.checkAlgorithm(algorithm);
            _this.checkKeyGenParams(algorithm);
            _this.checkKeyGenUsages(keyUsages);
            resolve(undefined);
        });
    };
    Aes.exportKey = function (format, key) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.checkKey(key, _this.ALG_NAME);
            _this.checkFormat(format, key.type);
            resolve(undefined);
        });
    };
    Aes.importKey = function (format, keyData, algorithm, extractable, keyUsages) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.checkAlgorithm(algorithm);
            _this.checkFormat(format);
            if (!(format.toLowerCase() === "raw" || format.toLowerCase() === "jwk"))
                throw new error_1.CryptoKeyError(error_1.CryptoKeyError.ALLOWED_FORMAT, format, "'jwk' or 'raw'");
            _this.checkKeyGenUsages(keyUsages);
            resolve(undefined);
        });
    };
    Aes.ALG_NAME = "";
    Aes.KEY_USAGES = [];
    return Aes;
}(base_1.BaseCrypto));
exports.Aes = Aes;
var AesAlgorithmError = (function (_super) {
    __extends(AesAlgorithmError, _super);
    function AesAlgorithmError() {
        _super.apply(this, arguments);
        this.code = 8;
    }
    return AesAlgorithmError;
}(error_1.AlgorithmError));
exports.AesAlgorithmError = AesAlgorithmError;
var AesEncrypt = (function (_super) {
    __extends(AesEncrypt, _super);
    function AesEncrypt() {
        _super.apply(this, arguments);
    }
    AesEncrypt.encrypt = function (algorithm, key, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.checkAlgorithmParams(algorithm);
            _this.checkKey(key, _this.ALG_NAME, "secret", "encrypt");
            resolve(undefined);
        });
    };
    AesEncrypt.decrypt = function (algorithm, key, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.checkAlgorithmParams(algorithm);
            _this.checkKey(key, _this.ALG_NAME, "secret", "decrypt");
            resolve(undefined);
        });
    };
    AesEncrypt.wrapKey = function (format, key, wrappingKey, wrapAlgorithm) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.checkAlgorithmParams(wrapAlgorithm);
            _this.checkKey(wrappingKey, _this.ALG_NAME, "secret", "wrapKey");
            _this.checkWrappedKey(key);
            _this.checkFormat(format, key.type);
            resolve(undefined);
        });
    };
    AesEncrypt.unwrapKey = function (format, wrappedKey, unwrappingKey, unwrapAlgorithm, unwrappedKeyAlgorithm, extractable, keyUsages) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.checkAlgorithmParams(unwrapAlgorithm);
            _this.checkKey(unwrappingKey, _this.ALG_NAME, "secret", "unwrapKey");
            _this.checkFormat(format);
            // TODO check unwrappedKeyAlgorithm
            // TODO check keyUSages
            resolve(undefined);
        });
    };
    AesEncrypt.KEY_USAGES = ["encrypt", "decrypt", "wrapKey", "unwrapKey"];
    return AesEncrypt;
}(Aes));
exports.AesEncrypt = AesEncrypt;
var AesCBC = (function (_super) {
    __extends(AesCBC, _super);
    function AesCBC() {
        _super.apply(this, arguments);
    }
    AesCBC.checkAlgorithmParams = function (alg) {
        this.checkAlgorithm(alg);
        if (!alg.iv)
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_REQUIRED, "iv");
        if (!ArrayBuffer.isView(alg.iv))
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_TYPE, "iv", "ArrayBufferView");
        if (alg.iv.byteLength !== 16)
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_VALUE, "iv", "ArrayBufferView with size 16");
    };
    AesCBC.ALG_NAME = alg_1.AlgorithmNames.AesCBC;
    return AesCBC;
}(AesEncrypt));
exports.AesCBC = AesCBC;
var AesCTR = (function (_super) {
    __extends(AesCTR, _super);
    function AesCTR() {
        _super.apply(this, arguments);
    }
    AesCTR.checkAlgorithmParams = function (alg) {
        this.checkAlgorithm(alg);
        if (!(alg.counter && ArrayBuffer.isView(alg.counter)))
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_TYPE, "counter", "ArrayBufferView");
        if (alg.counter.byteLength !== 16)
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_VALUE, "counter", "ArrayBufferView with size 16");
        if (!(alg.length > 0 && alg.length <= 128))
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_VALUE, "length", "number [1-128]");
    };
    AesCTR.ALG_NAME = alg_1.AlgorithmNames.AesCTR;
    return AesCTR;
}(AesEncrypt));
exports.AesCTR = AesCTR;
var AesGCM = (function (_super) {
    __extends(AesGCM, _super);
    function AesGCM() {
        _super.apply(this, arguments);
    }
    AesGCM.checkAlgorithmParams = function (alg) {
        this.checkAlgorithm(alg);
        if (alg.additionalData)
            if (!ArrayBuffer.isView(alg.additionalData))
                throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_TYPE, "additionalData", "ArrayBufferView");
        if (!alg.iv)
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_REQUIRED, "iv");
        if (!ArrayBuffer.isView(alg.iv))
            throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_TYPE, "iv", "ArrayBufferView");
        if (alg.tagLength)
            if (!(alg.tagLength >= 0 && alg.tagLength <= 128))
                throw new AesAlgorithmError(AesAlgorithmError.PARAM_WRONG_VALUE, "tagLength", "number [0-128]");
    };
    AesGCM.ALG_NAME = alg_1.AlgorithmNames.AesGCM;
    return AesGCM;
}(AesEncrypt));
exports.AesGCM = AesGCM;
