// pull-length-prefixed/lib/decode.js
'use strict';

var varint = require('varint');
var Reader = require('pull-reader');
var Buffer = require('safe-buffer').Buffer;
var pushable = require('pull-pushable');

exports.decode = decode;
exports.decodeFromReader = decodeFromReader;

var MSB = 0x80;
var isEndByte = function isEndByte(byte) {
  return !(byte & MSB);
};

function decode(opts) {
  var reader = new Reader();
  var p = pushable(function (err) {
    reader.abort(err);
  });

  return function (read) {
    reader(read);
    function next() {
      decodeFromReader(reader, opts, function (err, msg) {
        if (err) return p.end(err);

        p.push(msg);
        next();
      });
    }

    next();
    return p;
  };
}

function decodeFromReader(reader, opts, cb) {
  if (typeof opts === 'function') {
    cb = opts;
    opts = {};
  }

  opts = Object.assign({
    fixed: false,
    bytes: 4
  }, opts || {});

  if (opts.fixed) {
    readFixedMessage(reader, opts.bytes, cb);
  } else {
    readVarintMessage(reader, cb);
  }
}

function readFixedMessage(reader, byteLength, cb) {
  reader.read(byteLength, function (err, bytes) {
    if (err) {
      return cb(err);
    }

    var msgSize = bytes.readInt32BE(0);
    readMessage(reader, msgSize, cb);
  });
}

function readVarintMessage(reader, cb) {
  var rawMsgSize = [];
  if (rawMsgSize.length === 0) readByte();

  // 1. Read the varint
  function readByte() {
    reader.read(1, function (err, byte) {
      if (err) {
        return cb(err);
      }

      rawMsgSize.push(byte);

      if (byte && !isEndByte(byte[0])) {
        readByte();
        return;
      }

      var msgSize = varint.decode(Buffer.concat(rawMsgSize));
      readMessage(reader, msgSize, function (err, msg) {
        if (err) {
          return cb(err);
        }

        rawMsgSize = [];

        cb(null, msg);
      });
    });
  }
}

function readMessage(reader, size, cb) {
  reader.read(size, function (err, msg) {
    if (err) {
      return cb(err);
    }

    cb(null, msg);
  });
}
